### PR TITLE
Extend prometheus template with alertmanager and message-buffer

### PR DESCRIPTION
**Pull request by zgalor**
Prometheus Pod: prometheus behind oauth-proxy listening on 8443
alertmanager behind alert-buffer behind oauth-proxy listening on 9443
oauth-proxy validation using management token

Routes: 
prometheus - prometheus via prometheus-oauth-proxy
alerts - alert-buffer via alerts-oauth-proxy

cc @smarterclayton @simon3z

### CODE REVIEW

File: examples/prometheus/prometheus.yaml [hunk 1]
```diff
@@ -16,8 +16,17 @@ parameters:
   name: IMAGE_PROXY
   value: openshift/oauth-proxy:v1.0.0
 - description: The location of the prometheus image
+  # TODO: Change to official openshift build
   name: IMAGE_PROMETHEUS
-  value: openshift/prometheus:v2.0.0-dev
```
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: for some reason using this image creates an error:
`prometheus: error: unknown short flag '-s'`

File: examples/prometheus/prometheus.yaml [hunk 2]
```diff
@@ -126,9 +136,9 @@ objects:
 
         - name: prometheus
           args:
-          - --storage.tsdb.retention=6h
```
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: with the prom/prometheus image this had to change to `storage.local.retention`

**Comment by simon3z** [comment]
@smarterclayton ATM it seems we are using images that are not coming from the openshift ci (e.g. `prom/prometheus`, `prom/alertmanager`, `ilackarms/message-buffer`). Can you help us to have the proper images available?

**Comment by zgalor** [comment]
@moolitayer @ilackarms PTAL

File: examples/prometheus/prometheus.yaml (hunk 1 above)
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Because prometheus 2.0 doesn't support flags that only have a single dash.  Please revert this.

File: examples/prometheus/prometheus.yaml (hunk 2 above)
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: That's the old image, which is still 1.6.

File: examples/prometheus/prometheus.yaml [hunk 3]
```diff
@@ -291,3 +301,196 @@ objects:
         - source_labels: [__meta_kubernetes_service_name]
           action: replace
           target_label: kubernetes_name
+
+      alerting:
+        alertmanagers:
+        - scheme: http
+          static_configs:
+          - targets:
+            - "alertmanager:9093"
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: How are you defining authentication for this alert manager?

File: examples/prometheus/prometheus.yaml [hunk 4]
```diff
@@ -291,3 +301,196 @@ objects:
...
+- apiVersion: v1
+  kind: Secret
+  metadata:
+    name: alerts-proxy
+    namespace: "${NAMESPACE}"
+  stringData:
+    session_secret: "${SESSION_SECRET}="
+
+# Create an internal direct service to alertmanager app
+- apiVersion: v1
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: This means your app is not protected, which means anyone on the cluster can contact it (unless network policy is configured).  Why is having a separate alert manager deployment useful in the near term?

File: examples/prometheus/prometheus.yaml [hunk 5]
```diff
@@ -291,3 +301,196 @@ objects:
...
+          - -client-secret-file=/var/run/secrets/kubernetes.io/serviceaccount/token
+          - -cookie-secret-file=/etc/proxy/secrets/session_secret
+          - '-openshift-delegate-urls={"/": {"resource": "namespaces", "verb": "get", "name": "${NAMESPACE}"}}'
+          volumeMounts:
+          - mountPath: /etc/tls/private
+            name: alerts-tls
+          - mountPath: /etc/proxy/secrets
+            name: secrets
+
+        - name: message-buffer
```
Comment (ilackarms) [inline examples/prometheus/prometheus.yaml]: :+1:

File: examples/prometheus/prometheus.yaml [hunk 6]
```diff
@@ -291,3 +301,196 @@ objects:
...
+        # TODO:
+        group_by: []
+
+        # All the above attributes are inherited by all child routes and can
+        # overwritten on each.
+
+      receivers:
+      - name: message-buffer-wh
+        webhook_configs:
+        - url: http://localhost:9099/topics/alerts
```
Comment (ilackarms) [inline examples/prometheus/prometheus.yaml]: lgtm

File: examples/prometheus/prometheus.yaml [hunk 7]
```diff
@@ -291,3 +301,196 @@ objects:
...
+    name: alertmanager
+    namespace: "${NAMESPACE}"
+  data:
+    alertmanager.yml: |
+      global:
+
+      # The root route on which each incoming alert enters.
+      route:
+        # default route if none match
+        receiver: message-buffer-wh
```
Comment (moolitayer) [inline examples/prometheus/prometheus.yaml]: We need to have this root receiver, with a very prominent comment saying not to change it
(because manageiq won't work) but change hid child.
We also need to add a child route so that users will have a course of action for alerts.

This is an origin example so I don't know if you want to do it here or somewhere else.
If this is general example you might want to change this with something more generic(e.g smtp).

File: examples/prometheus/prometheus.yaml (hunk 4 above)
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: To my understanding, we interact with alertmanager/prometheus via oath_proxy. We need an oauth_proxy for each app, and alertmanager resides behind the message-buffer for persistency.
Prometheus needs to send alerts to alertmanager, so this is why we expose the service.
Are you concerned with the service allowing http (not https)?
Can we use the same oauth-proxy to route 2 different apps in the same pod?

File: examples/prometheus/prometheus.yaml (hunk 3 above)
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: I am not sure. should I use another oauth-proxy inside the alerts pod? or can the oauth-proxy support 2 apps? is there a simpler way?

File: examples/prometheus/prometheus.yaml (hunk 4 above)
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: The main reason for separation to 2 pods is that we might want to scale the prometheus pods.
The question is what is the best way to secure the alertmanager service?

File: examples/prometheus/prometheus.yaml [hunk 8]
```diff
@@ -133,24 +142,92 @@ objects:
           imagePullPolicy: IfNotPresent
           volumeMounts:
           - mountPath: /etc/prometheus
-            name: config-volume
+            name: prometheus-config
           - mountPath: /prometheus
-            name: data-volume
+            name: prometheus-data
+
+        # Deploy alertmanager behind prometheus-alert-buffer behind an oauth proxy
```
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: @smarterclayton This works, but I am not sure deploying behind 2 different proxies is the best practice. Is there a way for the oauth-proxy to support 2 incoming ports?

Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: You can support multiple upstreams if they don't overlap in URLs.  But I don't know why two proxies is an issue - you can target services to any pod you want, in any port you want.

**Comment by zgalor** [comment]
@smarterclayton Is there anything else missing in the PR apart from official builds that we will soon have?

**Comment by smarterclayton** [comment]
Maybe.  See https://github.com/openshift/origin/pull/15389
<https://github.com/openshift/origin/pull/15389#issuecomment-316939938> for
changes coming in 1.7.  I picked changes out and have a core test.

Will try and review today

On Jul 21, 2017, at 4:24 AM, Zohar Galor <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> Is there anything else
missing in the PR apart from official builds that we will soon have?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/15078#issuecomment-316938383>,
or mute
the thread

**Comment by smarterclayton** [comment]
Core templates were updated, please rebase on top of my configs and templates

**Comment by zgalor** [comment]
rebased and added an alert rule example

**Comment by moolitayer** [comment]
I think we should have a node selector for the Prometheus pods.

The world view I have (that might be wrong) is that we have infra nodes for all hosted services and compute nodes for users workloads, So i'm thinking:
```
nodeSelector:
  region: infra
```

Two other quick points outside the scope of this PR, maybe someone can clear up if these are really problems and we should create issues for them:
- I've noticed hawkular and logging(and not the fluentd collectors) don't respect the infra/compute nodes
distinctions
- @yaakov and @cben did a through research today and they noticed the router is deployed using a replication controller with an infra node selector - this should mean that we can get a situation where we have two routers on one infra node and none in another?
(correct me if I'm wrong)

**Comment by smarterclayton** [comment]
Operators have to run nodes, and so at minimum you want operational metrics from all nodes.

Also, the distinction between user and operational workloads is not as strong as you'd think in most setups, so the line between CPU and not is blurry.  I think a node selector might be reasonable at some point, but it's easy for someone to add it themselves.

**Comment by smarterclayton** [comment]
Router and registry are deployed into projects which have a default node selector of type=infra.  We almost always run multiple copies of those, and they will only run on infra nodes.

**Comment by smarterclayton** [comment]
If you're referring to placement - the node selector for prometheus should
be chosen by the namespace it is created in, and most infra components
should run on infra nodes for security and isolation reasons.  However,
given its size I'm not positive that prometheus is automatically out of the
box a good fit for infra by default, and may be something that is applied
later.  We don't force the use of infra nodes in all possible clusters, we
just recommend it as a useful default.

On Jul 31, 2017, at 6:58 AM, Mooli Tayer <notifications@github.com> wrote:

I think we should have a node selector for the Prometheus pods.

The world view I have (that might be wrong) is that we have infra nodes for
all hosted services and compute nodes for users workloads, So i'm thinking:

nodeSelector:
  region: infra

Two other quick points outside the scope of this PR, maybe someone can
clear up if these are really problems and we should create issues for them:

   - I've noticed hawkular and logging(and not the fluentd collectors)
   don't respect the infra/compute nodes
   distinctions
   - @yaakov <https://github.com/yaakov> and @cben <https://github.com/cben>
   did a through research today and they noticed the router is deployed using
   a replication controller with an infra node selector - this should mean
   that we can get a situation where we have two routers on one infra node and
   none in another?
   (correct me if I'm wrong)

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/15078#issuecomment-319035667>,
or mute
the thread

File: examples/prometheus/prometheus.yaml [hunk 9]
```diff
@@ -5,9 +5,9 @@ metadata:
   annotations:
     "openshift.io/display-name": Prometheus
     description: |
-      A monitoring solution for an OpenShift cluster - collect and gather metrics from nodes, services, and the infrastructure. This is a tech preview feature.
+      A monitoring solution for an OpenShift cluster - collect and gather metrics and alerts from nodes, services, and the infrastructure. This is a tech preview feature.
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Don't remove the tech preview comment.  This will still be tech preview.

File: examples/prometheus/prometheus.yaml [hunk 10]
```diff
@@ -18,8 +18,19 @@ parameters:
...
   name: IMAGE_PROMETHEUS
   value: openshift/prometheus:v2.0.0-dev
+- description: The location of the alertmanager image
+  name: IMAGE_ALERTMANAGER
+  # TODO: Change to official openshift build
+  value: prom/alertmanager
+- description: The location of alert-buffer image
+  name: IMAGE_ALERT_BUFFER
+  # TODO: change to official openshift build
+  value: ilackarms/message-buffer
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: This is fine for now unless you already have builds.

File: examples/prometheus/prometheus.yaml [hunk 11]
```diff
@@ -18,8 +18,19 @@ parameters:
...
 - description: The session secret for the proxy
-  name: SESSION_SECRET
+  name: PROMETEHUS_SESSION_SECRET
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: This is a typo.  Also, why are you splitting them if they are deployed together?

File: examples/prometheus/prometheus.yaml [hunk 12]
```diff
@@ -18,8 +18,19 @@ parameters:
 - description: The location of the prometheus image
   name: IMAGE_PROMETHEUS
   value: openshift/prometheus:v2.0.0-dev
+- description: The location of the alertmanager image
+  name: IMAGE_ALERTMANAGER
+  # TODO: Change to official openshift build
+  value: prom/alertmanager
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Should change to the origin one we published.

File: examples/prometheus/prometheus.yaml [hunk 13]
```diff
@@ -77,8 +88,7 @@ objects:
     name: prometheus-proxy
     namespace: "${NAMESPACE}"
   stringData:
-    session_secret: "${SESSION_SECRET}="
-# Deploy Prometheus behind an oauth proxy
+    session_secret: "${PROMETHEUS_SESSION_SECRET}="
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Would prefer to just reuse the session secret.

File: examples/prometheus/prometheus.yaml [hunk 14]
```diff
@@ -286,3 +377,76 @@ objects:
...
+  metadata:
+    name: alerts
+    namespace: "${NAMESPACE}"
+  spec:
+    to:
+      name: alerts
+    tls:
+      termination: Reencrypt
+
+# Create a service to access alerts vi auth-proxy
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Typo

File: examples/prometheus/prometheus.yaml [hunk 15]
```diff
@@ -286,3 +377,76 @@ objects:
...
+- apiVersion: v1
+  kind: Service
+  metadata:
+    annotations:
+      service.alpha.openshift.io/serving-cert-secret-name: alerts-tls
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: scope this secret to prometheus-alerts-tls

File: examples/prometheus/prometheus.yaml [hunk 16]
```diff
@@ -286,3 +377,76 @@ objects:
...
+  metadata:
+    name: alerts-proxy
+    namespace: "${NAMESPACE}"
+  stringData:
+    session_secret: "${ALERTS_SESSION_SECRET}="
+
+- apiVersion: v1
+  kind: ConfigMap
+  metadata:
+    name: alertmanager
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: scope this to prometheus-alerts

File: examples/prometheus/prometheus.yaml (hunk 12 above)
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: @smarterclayton where is it?

**Comment by smarterclayton** [comment]
https://hub.docker.com/r/openshift/prometheus-alertmanager/

File: examples/prometheus/prometheus.yaml [hunk 17]
```diff
@@ -139,6 +147,58 @@ objects:
...
+        - name: alerts-oauth-proxy
+          image: ${IMAGE_PROXY}
+          imagePullPolicy: IfNotPresent
+          ports:
+          - containerPort: 9443
+            name: web
+          args:
+          - -provider=openshift
+          - -https-address=:9443
+          - -http-address=:4190
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Why is this necessary

File: examples/prometheus/prometheus.yaml [hunk 18]
```diff
@@ -99,7 +106,8 @@ objects:
       spec:
         serviceAccountName: prometheus
         containers:
-        - name: oauth-proxy
+        # Deploy Prometheus behind an oauth proxy
+        - name: prometheus-oauth-proxy
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Call this prom-proxy since we're going to be typing it all the time, and the other container "prom", and then "alert-proxy" and "alert".

File: examples/prometheus/prometheus.yaml [hunk 19]
```diff
@@ -286,3 +374,76 @@ objects:
...
+
+      alerting:
+        alertmanagers:
+        - scheme: http
+          static_configs:
+          - targets:
+            - "localhost:9093"
+
+# Create a fully end-to-end TLS connection to the proxy
+- apiVersion: route.openshift.io/v1
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Who accesses alerts from outside the cluster?  Miq?

Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Comment here about this being exposed for external use if so (clients who wish to grab alter status).

File: examples/prometheus/prometheus.yaml (hunk 18 above)
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: "alerts" (to be consistent with other names), I mean

File: examples/prometheus/prometheus.yaml [hunk 20]
```diff
@@ -139,6 +147,58 @@ objects:
           - mountPath: /prometheus
             name: prometheus-data
 
+        # Deploy alertmanager behind prometheus-alert-buffer behind an oauth proxy
+        - name: alerts-oauth-proxy
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: alerts-proxy

File: examples/prometheus/prometheus.yaml [hunk 21]
```diff
@@ -153,14 +213,42 @@ objects:
...
         - name: prometheus-data
           emptyDir: {}
+        - name: alertmanager-config
+          configMap:
+            defaultMode: 420
+            name: prometheus-alerts
+        - name: alerts-secrets
+          secret:
+            secretName: alerts-proxy
+        - name: prometheus-alerts-tls
```
Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Make the internal volume names consistent - don't change prefixes.

File: examples/prometheus/prometheus.yaml (hunk 17 above)
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: Since the 2 proxies are in the same pod, and oauth-proxy listens on http with default port 4180, I had to change the port on the second proxy so they don't collide.

File: examples/prometheus/prometheus.yaml (hunk 19 above)
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: yes MIQ. will add comment

File: examples/prometheus/prometheus.yaml (hunk 17 above)
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: I added a comment to explain that

Comment (smarterclayton) [inline examples/prometheus/prometheus.yaml]: Oh.  I think this is a bug in oauth-proxy - it shouldn't listen on HTTP unless told.  Fine for now, can you open an oauth-proxy issue and assign to me?

**Comment by smarterclayton** [comment]
LGTM, go ahead and squash and I'll tag.

File: examples/prometheus/prometheus.yaml (hunk 17 above)
Comment (zgalor) [inline examples/prometheus/prometheus.yaml]: opened. https://github.com/openshift/oauth-proxy/issues/7

**Comment by zgalor** [comment]
@smarterclayton squashed

**Comment by zgalor** [comment]
@smarterclayton all tests passed but `Submit Queue` that is waiting for you approval

**Comment by smarterclayton** [comment]
I'm trying this out locally and see a few things broken:

1. login is not working for alerts via oauth - it doesn't have the latest SAR rules.  I'll fix that
2. once the SAR rules are fixed for alerts, it's giving an oauth login error for alerts, because the service account has no oauth reference
3. there are no labels on the services and config maps, so you can't delete everything all at once.  Everything in this should have app=prometheus, with sub-labels for the various parts (like component=alerts and component=prometheus)

**Comment by smarterclayton** [comment]
I'll fix 1 and 2 locally but it looks like a lot of this hasn't been tested.

**Comment by smarterclayton** [comment]
Does alerts even have a UI component?  Getting 404 as soon as I punch through oauth when 1/2 are fixed?

**Comment by zgalor** [comment]
@smarterclayton you need to add '/topics/alerts'
It's configured in alertmanager configmap as the webhook url

**Comment by smarterclayton** [comment]
#15785 has the fixes for 1 and 2
